### PR TITLE
adds sync capabilities for HA

### PR DESCRIPTION
**Pull request by andrewpmartinez**
New PR, rebased on top of release-next due to fabric merge into the ziti repo.

### CODE REVIEW

File: common/pb/edge_ctrl_pb/edge_ctrl.proto [hunk 1]
```diff
@@ -113,44 +118,187 @@ message Service {
...
+
+  message Revocation {
+    string id = 1;
+    google.protobuf.Timestamp ExpiresAt = 2;
+  }
+
+  message Event {
+    Action action = 1;
+    uint64 index = 2;
+    bool isDistributedIndex = 3;
```
Comment (andrewpmartinez) [inline common/pb/edge_ctrl_pb/edge_ctrl.proto:222]: Pretty sure this can be removed now.

File: common/router_data_model.go [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	rdmContents.RouterDataModel.lastSaveIndex = &rdmContents.Index
+
+	return rdmContents.RouterDataModel, nil
+}
+
+type RouterDataModelCache interface {
```
Comment (andrewpmartinez) [inline common/router_data_model.go:95]: Might be left over.

File: common/router_data_model.go [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+
+func (rdm *RouterDataModel) GetEventChannel() <-chan *edge_ctrl_pb.DataState_Event {
+	if rdm.eventChannel == nil {
+		rdm.eventChannel = make(chan *edge_ctrl_pb.DataState_Event, rdm.bufferSize)
+	}
+
+	return rdm.eventChannel
+}
+
+func (rdm *RouterDataModel) broadcastEvent(event *edge_ctrl_pb.DataState_Event) {
```
Comment (andrewpmartinez) [inline common/router_data_model.go:113]: Change to `send*`

File: common/router_data_model.go [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+		rdm.PublicKeys.Set(model.PublicKey.Kid, model.PublicKey)
+	}
+
+	rdm.store(event)
+	rdm.broadcastEvent(event)
+}
+
+func (rdm *RouterDataModel) ApplyRevocationEvent(event *edge_ctrl_pb.DataState_Event, model *edge_ctrl_pb.DataState_Event_Revocation) {
+	if index, ok := rdm.EventCache.CurrentIndex(); ok && index >= event.Index {
+		// old event
```
Comment (andrewpmartinez) [inline common/router_data_model.go:323]: Consider adding panics if we get +1/+n indexes.

File: common/router_data_model.go [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+		rdm.Revocations.Set(model.Revocation.Id, model.Revocation)
+	}
+
+	rdm.store(event)
+	rdm.broadcastEvent(event)
+}
+
+func (rdm *RouterDataModel) GetDataState() *edge_ctrl_pb.DataState {
+	//the resulting events may be good to cache as the model should stagnate at some point as users stop making config changes
+	var events []*edge_ctrl_pb.DataState_Event
```
Comment (andrewpmartinez) [inline common/router_data_model.go:339]: This should probably lock for single reader.

File: common/router_data_model.go [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+	// Write the gzipped JSON data to the file
+	_, err = gz.Write(jsonBytes)
+
+	if err != nil {
+		pfxlog.Logger().WithError(err).Error("could not marshal router data model, could not compress and write")
+		return
+	}
+}
+
+type AccessPolicies struct {
```
Comment (andrewpmartinez) [inline common/router_data_model.go:457]: This needs doc.

File: controller/env/broker.go [hunk 7]
```diff
@@ -45,6 +46,9 @@ const (
 	EnrollmentExtendRouterVerifyRequestType = int32(edge_ctrl_pb.ContentType_EnrollmentExtendRouterVerifyRequestType)
 )
 
+type RouterSyncCache struct {
```
Comment (andrewpmartinez) [inline controller/env/broker.go:49]: Remove.

File: router/handler_edge_ctrl/hello.go [hunk 8]
```diff
@@ -79,25 +81,23 @@ func (h *helloHandler) HandleReceive(msg *channel.Message, ch channel.Channel) {
...
+				Version:       build.GetBuildInfo().Version(),
+				Listeners:     h.listeners,
 				Hostname:      h.hostname,
 				Protocols:     h.supportedProtocols,
 				ProtocolPorts: h.protocolPorts,
+				Data:          map[string]string{},
+			}
+
+			if serverHello.Data[common.DataRouterModel] == "true" {
+				clientHello.Data[common.DataRouterModel] = "true"
```
Comment (andrewpmartinez) [inline router/handler_edge_ctrl/hello.go:94]: Make headers. Not data.

Comment (andrewpmartinez) [reply router/handler_edge_ctrl/hello.go:94]: There are also capabilities added by Paul.

- The router sends a string array of capabilities.
- A bitmask that is sent from the controller

File: router/posture/os.go [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+import "github.com/openziti/ziti/common/pb/edge_ctrl_pb"
+
+type OsCheck struct {
+	*edge_ctrl_pb.DataState_PostureCheck
+	*edge_ctrl_pb.DataState_PostureCheck_OsList
+}
+
+func (m *OsCheck) Evaluate(state *Cache) *CheckError {
+	return nil
```
Comment (andrewpmartinez) [inline router/posture/os.go:11]: Missing.

File: common/event_cache.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+
+	callback()
+}
+
+func (cache *ForgetfulEventCache) Store(event *edge_ctrl_pb.DataState_Event, onSuccess OnStoreSuccess) error {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+
```
Comment (plorenz) [inline common/event_cache.go:52]: Should this allow stale events to be ignored and only error if we see the index jump? Are stale events likely to be a regular occurrence?

File: common/router_data_model.go (hunk 2 above)
Comment (andrewpmartinez) [reply common/router_data_model.go]: Removed

File: common/pb/edge_ctrl_pb/edge_ctrl.proto (hunk 1 above)
Comment (andrewpmartinez) [reply common/pb/edge_ctrl_pb/edge_ctrl.proto:222]: Removed

File: common/event_cache.go (hunk 10 above)
Comment (andrewpmartinez) [reply common/event_cache.go:52]: Yes. Taken care of.

File: router/handler_edge_ctrl/dataState.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+		pfxlog.Logger().WithError(err).Errorf("could not marshal data state event message")
+		return
+	}
+
+	model := common.NewReceiverRouterDataModel(state.RouterDataModelListerBufferSize)
+
+	for _, event := range newState.Events {
+		model.Apply(event)
+	}
+
```
Comment (plorenz) [inline router/handler_edge_ctrl/dataState.go:39]: state.routerDataModel should maybe be stored in an atomic if it's getting replaced at runtime. Unlikely to ever cause a problem

**Review by plorenz — APPROVED** [review]

File: router/handler_edge_ctrl/dataState.go (hunk 11 above)
Comment (andrewpmartinez) [reply router/handler_edge_ctrl/dataState.go:40]: Done.

**Pull request merged**